### PR TITLE
Emit Redirect Errors for Anonymous Logins

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch changes API server login behaviour so that redirect errors are emitted for anonymous logins when that login is attempting to access a migrated model.

This is necessary for a following patch that will redirect CMD consumers to the new controller.

## QA steps

- Bootstrap 3 controllers: `source`, `consume`, `dest`.
- `juju switch source && juju deploy mysql && juju offer mysql:db`.
- `juju switch consume && juju deploy wordpress && juju consume source:admin/default.mysql && juju add-relation wordpress mysql`.
- Wait for quiescence in this model.
- `juju switch dest && juju destroy-model default -y`.
- `juju switch source && juju migrate default dest`.
- `juju switch consume`
- Check the logs for the remote relations worker. Errors should indicate redirect instead of permission denied.

## Documentation changes

None

## Bug reference

N/A

### CODE REVIEW

File: worker/firewaller/firewaller.go [hunk 1]
```diff
@@ -429,11 +429,9 @@ func (fw *Firewaller) startUnit(unit *firewaller.Unit, machineTag names.MachineT
 	if err != nil {
 		return err
 	}
+
 	applicationTag := application.Tag()
 	unitTag := unit.Tag()
-	if err != nil {
```
Comment (manadart) [inline worker/firewaller/firewaller.go]: Drive-by removal of redundant check.

File: state/modelmigration.go [hunk 2]
```diff
@@ -837,38 +837,26 @@ func (st *State) LatestMigration() (ModelMigration, error) {
...
 }
 
-// CompletedMigration returns the most recent ModelMigration (if any)
-// for input model UUID that has been removed from the state after
-// it reached the DONE phase and caused the model to be relocated.
-func (st *State) CompletedMigration(modelUUID string) (ModelMigration, error) {
+// CompletedMigration returns the most recent migration for the
+// input model UUID if it reached the DONE phase and caused the model
+// to be relocated.
+func (st *State) CompletedMigrationForModel(modelUUID string) (ModelMigration, error) {
```
Comment (manadart) [inline state/modelmigration.go]: Some of the churn is due to this exchange of method names that better indicate their behaviour.

File: apiserver/stateauthenticator/auth.go [hunk 3]
```diff
@@ -221,10 +223,11 @@ func LoginRequest(req *http.Request) (params.LoginRequest, error) {
 	if _, err := names.ParseTag(tagPass[0]); err != nil {
 		return params.LoginRequest{}, errors.Trace(err)
 	}
+
 	return params.LoginRequest{
 		AuthTag:     tagPass[0],
 		Credentials: tagPass[1],
-		Macaroons:   httpbakery.RequestMacaroons(req),
+		Macaroons:   macaroons,
```
Comment (manadart) [inline apiserver/stateauthenticator/auth.go]: Drive-by. This variable was already materialised.

File: apiserver/admin.go [hunk 4]
```diff
@@ -246,58 +259,35 @@ func (a *admin) authenticate(req params.LoginRequest) (*authResult, error) {
...
-			controllerConn,
-			req.UserData,
-		)
-	} else if a.root.model == nil { // anonymous login to unknown model
-		// Hide the fact that the model does not exist
+		a.root.entity = authInfo.Entity
+		a.apiObserver.Login(authInfo.Entity.Tag(), a.root.model.ModelTag(), controllerConn, req.UserData)
+	} else if a.root.model == nil {
+		// Anonymous login to unknown model.
+		// Hide the fact that the model does not exist.
```
Comment (SimonRichardson) [inline apiserver/admin.go]: We should add login stats here at some point, to at least take note of how many times this happens etc.

**Comment by manadart** [comment]
$$merge$$
